### PR TITLE
Use TORCH_CUDA_ARCH_LIST to specify which CUDA architectures to build for

### PR DESCRIPTION
**Pull request by NicolasHug**
This PR uses the `TORCH_CUDA_ARCH_LIST` env variable to specify which virtual and real architectures to build for, instead of using the more verbose `NVCC_FLAGS` specification.

Closes https://github.com/pytorch/vision/issues/2377
Closes https://github.com/pytorch/vision/issues/2026

### CODE REVIEW

**Comment by NicolasHug** [comment]
@ezyang it seems that removing the `NVCC_FLAGS` leads to an ``IndexError`` in `pytorch`, as the packaging machines don't have GPUs. Would you know what an appropriate fix could be?

**Comment by ezyang** [comment]
Try setting `TORCH_CUDA_ARCH_LIST` in lieu of the manual NVCC flags. It's been a while since I wrote that comment, but I think my intent was that you still specify the arch list, just less low level.

**Comment by NicolasHug** [comment]
@seemethere I'm trying to replace `NVCC_FLAGS` with `TORCH_CUDA_ARCH_LIST` in torchvision, but this seems to lead to an error in pytorch's setup.py: https://app.circleci.com/pipelines/github/pytorch/vision/6799/workflows/557c1018-186a-40ea-8725-5302ef620374/jobs/455001

I'm not sure whether this is an issue with pytorch or if I'm doing something wrong. Would you mind taking a look? Thanks!

**Comment by ezyang** [comment]
Here's the code that's failing:

```
    # The default is sm_30 for CUDA 9.x and 10.x
    # First check for an env var (same as used by the main setup.py)
    # Can be one or more architectures, e.g. "6.1" or "3.5;5.2;6.0;6.1;7.0+PTX"
    # See cmake/Modules_CUDA_fix/upstream/FindCUDA/select_compute_arch.cmake
    _arch_list = os.environ.get('TORCH_CUDA_ARCH_LIST', None)

    # If not given, determine what's best for the GPU / CUDA version that can be found
    if not _arch_list:
        arch_list = []
        # the assumption is that the extension should run on any of the currently visible cards,
        # which could be of different types - therefore all archs for visible cards should be included
        for i in range(torch.cuda.device_count()):
            capability = torch.cuda.get_device_capability(i)
            supported_sm = [int(arch.split('_')[1])
                            for arch in torch.cuda.get_arch_list() if 'sm_' in arch]
            max_supported_sm = max((sm // 10, sm % 10) for sm in supported_sm)
            # Capability of the device may be higher than what's supported by the user's
            # NVCC, causing compilation error. User's NVCC is expected to match the one
            # used to build pytorch, so we use the maximum supported capability of pytorch
            # to clamp the capability.
            capability = min(max_supported_sm, capability)
            arch = f'{capability[0]}.{capability[1]}'
            if arch not in arch_list:
                arch_list.append(arch)
        arch_list = sorted(arch_list)
        arch_list[-1] += '+PTX'
```

This suggests that the environment variable isn't being passed through.  Add this env var to the conda meta.yaml here https://github.com/pytorch/vision/blob/978ba613518dd5b8d04c2a717e6e5ccf7fb172c3/packaging/torchvision/meta.yaml#L41

File: packaging/torchvision/meta.yaml [hunk 1]
```diff
@@ -39,6 +39,7 @@ build:
     - FORCE_CUDA
     - NVCC_FLAGS
```
Comment (fmassa) [inline packaging/torchvision/meta.yaml:40]: Maybe we can remove this flag now?

Comment (NicolasHug) [reply packaging/torchvision/meta.yaml:40]: I'm not sure we can remove it, as `NVCC_FLAGS` is still being used in the windows files:

https://github.com/pytorch/vision/blob/978ba613518dd5b8d04c2a717e6e5ccf7fb172c3/packaging/windows/cuda101.bat#L35-L40

Comment (fmassa) [reply packaging/torchvision/meta.yaml]: I wonder if we still need this given that PyTorch already set those in [cpp_extensions](https://github.com/peterjc123/pytorch/blob/65087dd1d4c1e503281cd676e5563e9e64b2a926/torch/utils/cpp_extension.py#L178-L183).

Maybe @peterjc123 could chime in?

Comment (NicolasHug) [reply packaging/torchvision/meta.yaml]: I removed it in 5055fa8 and it looks like the CI is still happy

Comment (fmassa) [reply packaging/torchvision/meta.yaml]: Yeah but that codepath might not be exercised on the PR CI, but only on the release CIs.

I'd love to get @peterjc123 to double-check this part before we merge this

Comment (peterjc123) [reply packaging/torchvision/meta.yaml]: You should do it in the script as well. `__CUDA_NO_HALF_OPERATORS__ ` and `-expt-relaxed-constexpr` are set in `torch/utils/cpp_extension.py` so I think it should be okay.

Comment (fmassa) [reply packaging/torchvision/meta.yaml]: Great, thanks for the confirmation @peterjc123 !

I've opened https://github.com/pytorch/vision/issues/3617 to track removing this file

**Review by fmassa — APPROVED** [review]
Thanks @NicolasHug

**Comment by xkszltl** [comment]
I wonder if we should have TORCH_CUDA_ARCH_LIST exported from pytorch to cmake configs, so that torch vision can build consistently with the pytorch it's relying on?

**Comment by fmassa** [comment]
@xksteven yes, I think this is a good idea and something I would love to see implemented in PyTorch. Can you create an issue in PyTorch to track this down?

**Pull request merged**